drivers/modem/wncm14a2a.c

```diff
@@ -0,0 +1,1927 @@
+}
+
+#define MDM_MANUFACTURER_LENGTH		10
+#define MDM_MODEL_LENGTH		16
+#define MDM_REVISION_LENGTH		64
+#define MDM_IMEI_LENGTH			16
+
+#define RSSI_TIMEOUT_SECS		30
+
+NET_BUF_POOL_DEFINE(mdm_recv_pool, MDM_RECV_MAX_BUF, MDM_RECV_BUF_SIZE,
+		    0, NULL);
+
+static u8_t mdm_recv_buf[MDM_MAX_DATA_LENGTH];
+
+/* RX thread structures */
+K_THREAD_STACK_DEFINE(wncm14a2a_rx_stack,
+		       CONFIG_MODEM_WNCM14A2A_RX_STACK_SIZE);
+struct k_thread wncm14a2a_rx_thread;
+
+/* RX thread work queue */
+K_THREAD_STACK_DEFINE(wncm14a2a_workq_stack,
+		      CONFIG_MODEM_WNCM14A2A_RX_WORKQ_STACK_SIZE);
+static struct k_work_q wncm14a2a_workq;
+
+struct wncm14a2a_socket {
+	struct net_context *context;
+	sa_family_t family;
+	enum net_sock_type type;
+	enum net_ip_protocol ip_proto;
+	struct sockaddr src;
+	struct sockaddr dst;
+	int dst_port;
+
+	int socket_id;
+	bool socket_reading;
+
+	/** semaphore */
+	struct k_sem sock_send_sem;
+
+	/** socket callbacks */
+	struct k_work recv_cb_work;
+	net_context_recv_cb_t recv_cb;
+	struct net_pkt *recv_pkt;
+	void *recv_user_data;
+};
+
+struct wncm14a2a_iface_ctx {
+	struct net_if *iface;
+	u8_t mac_addr[6];
+
+	/* GPIO PORT devices */
+	struct device *gpio_port_dev[MAX_MDM_CONTROL_PINS];
+
+	/* RX specific attributes */
+	struct mdm_receiver_context mdm_ctx;
+
+	/* socket data */
+	struct wncm14a2a_socket sockets[MDM_MAX_SOCKETS];
+	int last_socket_id;
+	int last_error;
+
+	/* semaphores */
+	struct k_sem response_sem;
+
+	/* RSSI work */
+	struct k_delayed_work rssi_query_work;
+
+	/* modem data */
+	char mdm_manufacturer[MDM_MANUFACTURER_LENGTH];
+	char mdm_model[MDM_MODEL_LENGTH];
+	char mdm_revision[MDM_REVISION_LENGTH];
+	char mdm_imei[MDM_IMEI_LENGTH];
+
+	/* modem state */
+	int ev_csps;
+	int ev_rrcstate;
+};
+
+struct cmd_handler {
+	const char *cmd;
+	u16_t cmd_len;
+	void (*func)(struct net_buf **buf, u16_t len);
+};
+
+static struct wncm14a2a_iface_ctx ictx;
+
+static void wncm14a2a_read_rx(struct net_buf **buf);
+
+/*** Verbose Debugging Functions ***/
+#if defined(ENABLE_VERBOSE_MODEM_RECV_HEXDUMP)
+static inline void _hexdump(const u8_t *packet, size_t length)
+{
+	char output[sizeof("xxxxyyyy xxxxyyyy")];
+	int n = 0, k = 0;
+	u8_t byte;
+
+	while (length--) {
+		if (n % 16 == 0) {
+			printk(" %08X ", n);
+		}
+
+		byte = *packet++;
+
+		printk("%02X ", byte);
+
+		if (byte < 0x20 || byte > 0x7f) {
+			output[k++] = '.';
+		} else {
+			output[k++] = byte;
+		}
+
+		n++;
+		if (n % 8 == 0) {
+			if (n % 16 == 0) {
+				output[k] = '\0';
+				printk(" [%s]\n", output);
+				k = 0;
+			} else {
+				printk(" ");
+			}
+		}
+	}
+
+	if (n % 16) {
+		int i;
+
+		output[k] = '\0';
+
+		for (i = 0; i < (16 - (n % 16)); i++) {
+			printk("   ");
+		}
+
+		if ((n % 16) < 8) {
+			printk(" "); /* one extra delimiter after 8 chars */
+		}
+
+		printk(" [%s]\n", output);
+	}
+}
+#else
+#define _hexdump(...)
+#endif
+
+static struct wncm14a2a_socket *socket_from_netctx(struct net_context *context)
```

Comment:
FYI: #8835 
If that patch is approved before this modem driver is merged, it might be possible to just use the net_context->offload_context field to store `the struct wncm14a2a_socket *`.

Reply:
Ah that would be great.